src/python/pants/backend/python/tasks/select_interpreter.py

```diff
@@ -69,11 +73,11 @@ def execute(self):
     with self.invalidated(python_tgts, fingerprint_strategy=fs) as invalidation_check:
```

Comment:
If the fs took a `PythonSetup` in its constructor and used `compatibility_or_constraints` this would be more correct (invalidate less). The path joining code below would be reverted in that case.

Reply:
I took a sort of similar approach originally: https://github.com/pantsbuild/pants/pull/7586#discussion_r276485378. Benjy and I decided it would be better to rely on the subsystem invalidation, but I think this is probably right.

Questions 1-3 from that PR still confuse me if. Will take a stab at this in the meantime.